clr/main.py

```diff
@@ -13,13 +28,40 @@ def main(argv=None):
     namespace = get_namespace(namespace_key)
     bound_args = namespace.parse_args(cmd_name, argv[2:])
 
-    # Call the command.
+    init_beeline(namespace_key)
+    trace = beeline.start_trace(context={
+        "name": cmd_name,
+        "username": getpass.getuser(),
+    })
+
     if hasattr(namespace.instance, "cmdinit"):
+        cmdinit_span = beeline.start_span({"name": "cmdinit"})
         namespace.instance.cmdinit()
-    result = namespace.command_callables[cmd_name](
-        *bound_args.args, **bound_args.kwargs
-    )
+        beeline.finish_span(cmdinit_span)
+
+    cmdrun_span = beeline.start_span({"name": "cmdrun"})
+    exit_code = 0
+    result = None
+    try:
+        result = namespace.command_callables[cmd_name](
+            *bound_args.args, **bound_args.kwargs
+        )
+    except Exception as e:
+        print('Error running command', e)
+        beeline.add_context_field('raised_exception', True)
+        exit_code = 1
+
+    beeline.finish_span(cmdrun_span)
+
     # Support exit codes. This supports int and bool return values.
-    if isinstance(result, (int, bool)):
-        # print(f"Result: {result}", file=sys.stderr)
-        sys.exit(result)
+    if result is not None and isinstance(result, (int, bool)):
+        exit_code = result
+
+    beeline.add_context_field('exit_code', exit_code)
+    beeline.finish_trace(trace)
+    beeline.close()
+
+    # only exit if we have a non-zero code otherwise we want python to handle
+    # exit codes
+    if exit_code != 0:
+        sys.exit(exit_code)
```

Comment:
nit: since we're exiting anyway you could don't need the if check. `sys.exit(0)` is ok

Reply:
Without this all the tests error due to us explicitly calling `sys.exit`.

Reply:
```
def test_argtest(capsys):
    def check(args, expected):
        with pytest.raises(SystemExit) as e:
            clr.main(["clr", "argtest"] + args)
            captured = capsys.readouterr()
            assert expected in captured.out
        assert e.value.code == 0
```

if you want. i don't feel strongly

Reply:
ahh, good to know - I was not sure how to handle it

Reply:
It's very confusing to me: 
```
===================================================================== test session starts =====================================================================
platform darwin -- Python 3.8.9, pytest-6.2.4, py-1.10.0, pluggy-1.0.0.dev0
rootdir: /Users/jon.carl/code/clr
collected 2 items

tests/test_clr.py .F                                                                                                                                    [100%]

========================================================================== FAILURES ===========================================================================
________________________________________________________________________ test_argtest2 ________________________________________________________________________

capsys = <_pytest.capture.CaptureFixture object at 0x111c7c730>

    def test_argtest2(capsys):
        def check(args, expected):
            clr.main(["clr", "argtest2"] + args)
            captured = capsys.readouterr()
            assert expected in captured.out

        def check_failure(args, expected):
            with pytest.raises(SystemExit):
                clr.main(["clr", "argtest2"] + args)
            captured = capsys.readouterr()
            assert expected in captured.err

>       check(["1", "2"], "a=1 b=2 c=() d=4 e=None f=False g=")

tests/test_clr.py:54:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/test_clr.py:44: in check
    clr.main(["clr", "argtest2"] + args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

argv = ['clr', 'argtest2', '1', '2']

    def main(argv=None):
        if not argv:
            argv = sys.argv
        query = "system:help"
        if len(argv) > 1:
            query = argv[1]

        namespace_key, cmd_name = resolve_command(query)
        namespace = get_namespace(namespace_key)
        bound_args = namespace.parse_args(cmd_name, argv[2:])

        init_beeline(namespace_key)
        trace = beeline.start_trace(context={
            "name": cmd_name,
            "username": getpass.getuser(),
            "hostname": socket.gethostname(),
        })

        if hasattr(namespace.instance, "cmdinit"):
            cmdinit_span = beeline.start_span({"name": "cmdinit"})
            namespace.instance.cmdinit()
            beeline.finish_span(cmdinit_span)

        cmdrun_span = beeline.start_span({"name": "cmdrun"})
        exit_code = 0
        result = None
        try:
            result = namespace.command_callables[cmd_name](
                *bound_args.args, **bound_args.kwargs
            )
        except Exception as e:
            print('Error running command: %s', e, file=sys.stderr)
            beeline.add_context_field('raised_exception', True)
            exit_code = 1

        beeline.finish_span(cmdrun_span)

        # Support exit codes. This supports int and bool return values.
        if result is not None and isinstance(result, (int, bool)):
            exit_code = result

        beeline.add_context_field('exit_code', exit_code)
        beeline.finish_trace(trace)
        beeline.close()

>       sys.exit(exit_code)
E       SystemExit: 0

clr/main.py:66: SystemExit
-------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------
a=1 b=2 c=() d=4 e=None f=False g=
=================================================================== short test summary info ===================================================================
FAILED tests/test_clr.py::test_argtest2 - SystemExit: 0
================================================================= 1 failed, 1 passed in 0.24s =================================================================
```